testctrl/svc/types/component.go

```diff
@@ -0,0 +1,120 @@
+
+import (
+	"fmt"
+
+	"github.com/google/uuid"
+	pb "github.com/grpc/grpc/testctrl/genproto/testctrl/svc"
+)
+
+// Component represents a dependency for a session. Benchmarks tend to have three components: a
+// driver, server and client. Each component can have many identical replicas.
+type Component struct {
+	name           string
+	containerImage string
+	kind           ComponentKind
+	replicas       int32
+	env            map[string]string
+}
+
+// NewComponent creates a Component instance, given a container image, kind and replicas.
+//
+// The container image string should contain the path to a docker image in a registry, versioned by
+// an explicit tag or hash. Replicas should be 1 for drivers and servers. For clients, they may be
+// any positive integer.
+//
+// For example:
+//
+//     // Container uses "nginx" image with explicit tag "latest" from docker hub
+//     a := NewComponent("nginx:latest", ServerComponent, 1)
+//
+//     // Container uses "java_worker" image with explicit tag "v3.2.2" from GCR with 2 replicas
+//     b := NewComponent("gcr.io/grpc-testing/java_worker:v3.2.2", ClientComponent, 2)
+//
+//     // Container uses "driver" image with hash from GCR
+//     c := NewComponent("gcr.io/grpc-testing/driver@sha256:e4ff8efd7eb62d3a3bb0990f6ff1e9df20da24ebf908d08f49cb83422a232862", DriverComponent, 1)
+func NewComponent(containerImage string, kind ComponentKind, replicas int32) *Component {
+	return &Component{
+		name:           uuid.New().String(),
+		containerImage: containerImage,
+		kind:           kind,
+		replicas:       replicas,
+	}
+}
+
+// Name returns a string that uniquely identifies a component. This name is shared by all replicas,
+// but it is not shared by any other component in the same session or any identical component in
+// another session.
+func (c *Component) Name() string {
+	return c.name
+}
+
+// ResourceName returns the name the component prefixed with `components/`. This value should be
+// displayed to a consumer of the API.
+func (c *Component) ResourceName() string {
+	return fmt.Sprintf("components/%s", c.Name())
+}
+
+// ContainerImage returns the name of a docker image and tag which contains the required component.
+func (c *Component) ContainerImage() string {
+	return c.containerImage
+}
+
+// Kind returns the type of component.
+func (c *Component) Kind() ComponentKind {
+	return c.kind
+}
+
+// Replicas returns the number of replicas of this component.
+func (c *Component) Replicas() int32 {
+	return c.replicas
+}
+
+// SetEnv stores an environment variable that must be set on the component. Setting one after a
+// component is running will have no effect.
+func (c *Component) SetEnv(key, value string) {
+	c.env[key] = value
+}
+
+// ComponentKind specifies the type of component the test requires.
+type ComponentKind int32
+
+const (
+	// DriverComponent is a test component that orchestrates workers, such as clients and servers.
+	DriverComponent ComponentKind = iota
```

Comment:
I would make this "iota + 1" so that the numeric values in go match the ones in the proto.

Reply:
I fixed it just now. I used the blank identifier `_ ComponentKind = iota` for consistency. This also sets it to start at 1 by ignoring the first value. I had already used this on the `EventKind` and figured I'd keep them the same. Good to know you can `iota + 1`. I did not know that was possible.